syntax = "proto3";

package relationalai.lqp.v1;

message Declaration {
  oneof declaration_type {
    Def def = 1;
    Loop loop = 2;
  }
}

message Def {
  RelationId name = 1;
  Abstraction body = 2;
  repeated Attribute attrs = 3;
}

message Loop {
  LoopIndex temporal_var = 1;
  repeated Def inits = 2;
  repeated Declaration body = 3;
}

// Abstraction and related types
message Abstraction {
  repeated Var vars = 1;
  Formula value = 2;
}

// Formula variants
message Formula {
  oneof formula_type {
    Exists exists = 1;
    Reduce reduce = 2;
    Conjunction conjunction = 3;
    Disjunction disjunction = 4;
    Not not = 5;
    FFI ffi = 6;
    Atom atom = 7;
    Pragma pragma = 8;
    Primitive primitive = 9;
    True true_val = 10;
    False false_val = 11;
    RelAtom rel_atom = 12;
  }
}

message Exists {
  repeated Var vars = 1;
  Formula value = 2;
}

message Reduce {
  Abstraction op = 1;
  Abstraction body = 2;
  repeated Term terms = 3;
}

message Conjunction {
  repeated Formula args = 1;
}

message Disjunction {
  repeated Formula args = 1;
}

message Not {
  Formula arg = 1;
}

message FFI {
  string name = 1;
  repeated Abstraction args = 2;
  repeated Term terms = 3;
}

message Atom {
  RelationId name = 1;
  repeated Term terms = 2;
}

message Pragma {
  string name = 1;
  repeated Term terms = 2;
}

message Primitive {
  string name = 1;
  repeated Term terms = 2;
}

message True {}
message False {}

message RelAtom {
  RelationSig sig = 1;
  repeated Term terms = 2;
}

message Term {
  oneof term_type {
    Var var = 1;
    Constant constant = 2;
  }
}

message Var {
  string name = 1; // Symbol as the variable's identifier
  RelType type = 2; // Reference to the type
}

message Constant {
  PrimitiveValue value = 1; // Actual value instance
}

message Attribute {
  string name = 1;
  repeated Constant args = 2;
}

message RelationId {
  fixed64 id_low = 1; // Lower 64 bits of UInt128
  fixed64 id_high = 2; // Upper 64 bits of UInt128
}

message LoopIndex {
  string name = 1; // Assuming LoopIndex is a named identifier
}

message RelType {
  oneof rel_type {
    PrimitiveType primitive_type = 1;
    RelValueType value_type = 2;
  }
}

enum PrimitiveType {
  PRIMITIVE_TYPE_UNSPECIFIED = 0;
  PRIMITIVE_TYPE_STRING = 1;
  PRIMITIVE_TYPE_INT = 2;
  PRIMITIVE_TYPE_FLOAT = 3;
  PRIMITIVE_TYPE_UINT128 = 4;
}

enum RelValueType {
  PRIMITIVE_TYPE_DECIMAL = 1;

  PRIMITIVE_TYPE_DATE = 2;
  PRIMITIVE_TYPE_DATETIME = 3;

  PRIMITIVE_TYPE_NANOSECOND = 4;
  PRIMITIVE_TYPE_MICROSECOND = 5;
  PRIMITIVE_TYPE_MILLISECOND = 6;
  PRIMITIVE_TYPE_SECOND = 7;
  PRIMITIVE_TYPE_MINUTE = 8;
  PRIMITIVE_TYPE_HOUR = 9;
  PRIMITIVE_TYPE_DAY = 10;
  PRIMITIVE_TYPE_WEEK = 11;
  PRIMITIVE_TYPE_MONTH = 12;
  PRIMITIVE_TYPE_YEAR = 13;
}

message PrimitiveValue {
  oneof value {
    string string_value = 1;
    int64 int_value = 2;
    double float_value = 3;
    UInt128 uint128_value = 4;
  }
}

message UInt128 {
  fixed64 hash_low = 1; // Lower 64 bits of UInt128
  fixed64 hash_high = 2; // Upper 64 bits of UInt128
}

message RelationSig {
  string name = 1; // Name of the relation
  repeated RelType types = 2; // Attributes of the relation
}
